plutus-metatheory/src/RawU.lagda.md

```diff
@@ -0,0 +1,308 @@
+
+This module provides the interface between the Haskell code and the Agda code.
+It replicates the Haskell representation of Raw. In particular, for term constants,
+we need the following extension to be able to replicate the Haskell representation
+of Universes.
+
+```
+{-# OPTIONS --type-in-type #-}
+
+module RawU where
+```
+
+## Imports
+
+```
+open import Function using (_∘_)
+open import Data.Nat using (ℕ)
+open import Data.Bool using (true;false)
+open import Data.Integer using (ℤ)
+open import Data.String using (String)
+open import Data.Bool using (Bool)
+open import Relation.Binary using (DecidableEquality)
+open import Relation.Binary.PropositionalEquality using (_≡_;refl;sym;cong;cong₂)
+open import Relation.Nullary using (yes;no;¬_)
+open import Data.Unit using (⊤;tt)
+open import Data.Product using (Σ;proj₁;proj₂) renaming (_,_ to _,,_)
+
+open import Utils using (ByteString;DATA;List;_×_;_,_)
+open import Builtin using (Builtin;equals)
+open Builtin.Builtin
+
+{-# FOREIGN GHC {-# LANGUAGE GADTs #-} #-}
+{-# FOREIGN GHC import PlutusCore #-}
+{-# FOREIGN GHC import Raw #-}
+```
+
+## A (Haskell) Universe for types
+
+The following tags use type-in-type, and map more directly to the Haskell representation
+of the default universe.
+
+Tags are indexed by the real type they represent.
+The `Esc` datatype is used in the Haskell implementation to "escape" any kind into Type.
+For constants, we only care about kind *, but we need it to match the Haskell implementation.
+```
+data Esc (a : Set) : Set where
+{-# INJECTIVE Esc #-}
+{-# COMPILE GHC Esc = data Esc () #-}
+
+data Tag : Set → Set where
+  integer    : Tag (Esc ℤ)
+  bytestring : Tag (Esc ByteString)
+  string     : Tag (Esc String)
+  bool       : Tag (Esc Bool)
+  unit       : Tag (Esc ⊤)
+  pdata      : Tag (Esc DATA)
+  pair       : ∀{A B} → Tag (Esc A) → Tag (Esc B) → Tag (Esc (A × B))
+  list       : ∀{A} → Tag (Esc A) → Tag (Esc (List A))
+
+{-# FOREIGN GHC type Tag = DefaultUni #-}
+{-# FOREIGN GHC pattern TagInt        = DefaultUniInteger  #-}
+{-# FOREIGN GHC pattern TagBS         = DefaultUniByteString #-}
```

Comment:
... although looking at that I guess it's easiest to just replicate the implementation's nonsense indeed.